common/src/main/java/com/wynntils/services/itemfilter/SearchQuery.java

```diff
@@ -0,0 +1,162 @@
+                // We still want to highlight the filter keyword though, that's why we handle the empty value here.
+                if (valueString.isEmpty()) continue;
+                ErrorOr<? extends ItemFilter> filterOrError =
+                        factoryOrError.getValue().create(valueString);
+
+                // If the filter value is invalid, mark the value as ignored and continue to the next token
+                if (filterOrError.hasError()) {
+                    ignoredCharIndices.addAll(IntStream.rangeClosed(
+                                    tokenStartIndex + filterString.length() + 1, tokenStartIndex + token.length())
+                            .boxed()
+                            .toList());
+                    errors.add(filterOrError.getError());
+                    continue;
+                }
+
+                // The filter value is valid, add the filter to the list
+                itemFilters.add(filterOrError.getValue());
+            } else if (!token.isEmpty()) {
+                // The token is not a filter, add it to the list of plain text tokens
+                plainTextTokens.add(token);
+            }
+        }
+
+        return new SearchQuery(
+                queryString, itemFilters, ignoredCharIndices, validFilterCharIndices, errors, plainTextTokens);
+    }
+
+    /**
+     * Checks if the given item matches all filters. Tokens that are not filters in the search query are ignored. If no
+     * filters are present, this method always returns true.
+     *
+     * @param wynnItem the item to check
+     * @return true if the item matches all filters, false otherwise
+     */
+    public boolean filterMatches(WynnItem wynnItem) {
+        return itemFilters.stream().allMatch(o -> o.matches(wynnItem));
+    }
+
+    /**
+     * Checks if the given item name contains the concatenated plain text tokens of the search query. The filter tokens
+     * in the search query are ignored. If there are no plain text tokens, this method always returns true.
+     *
+     * @param itemName the name to check
+     * @return true if the name contains the concatenated plain text tokens, false otherwise
+     */
+    public boolean itemNameMatches(String itemName) {
+        return plainTextTokens.isEmpty()
+                || itemName.toLowerCase(Locale.ROOT)
+                        .contains(String.join(" ", plainTextTokens).toLowerCase(Locale.ROOT));
+    }
+
+    /**
+     * Checks if the query contains no valid filters or plain text tokens.
+     *
+     * @return true if the query contains no valid filters or plain text tokens.
+     */
+    public boolean isEmpty() {
+        return itemFilters.isEmpty() && plainTextTokens.isEmpty();
+    }
+
+    public String getQueryString() {
+        return queryString;
+    }
+
+    public List<Integer> getIgnoredCharIndices() {
+        return Collections.unmodifiableList(ignoredCharIndices);
```

Comment:
I know wrapping lists to be returned in a read-only container is a recommended best practice, but we don't do it elsewhere in the code base. Our assumption has been (at least that's how I've seen it) that we trust the users of the APIs we provide to know that they should not modify collections, or make a copy if they need to. This makes the code slightly more fragile, but lets us get rid of a lot of boilerplate, and be able to treat returns of collections like returns of any other variable.

@kristofbolyai Should we start doing this? I have been pondering it every time I see that yellow marker from IntelliJ. I see the point of getting the extra safety, but I really dislike having all this extra boilerplate. On the other hand, IntelliJ can probably help with writing it, so it is mostly a matter of it being in the way when reading the code.

Reply:
I did it in some places where I saw it fit. But not everywhere, which is not good. To be honest, if we do this, we should also make models/services/handlers/managers return optionals..

Reply:
> if we do this, we should also make models/services/handlers/managers return optionals..

That is a much more and serious problem we have, since that has caused a lot of real-world problems with NPEs. The modifying of returned collections is more of a theoretic problem.